### DISCUSSION TITLE
Property dependency on another property's value

### DISCUSSION THREAD

**Discussion by Nahuel92** [micronaut-projects/micronaut-core#11119] [Q&A] [answered]
### Issue description

Hi there!

> [!NOTE]
> I wasn't sure how to tag this, feel free to re-tag if the current one is wrong. Thanks.

I have the following use case:

When an app is running locally and a `debug` flag is set to true, I need Micronaut to enforce read-only transactions. But when the app is running in any deployable env., I need Micronaut to remove that restriction (debug mode is intended for developers only on their machines).

The app is using a custom env. variable called "environment" (we use yaml properties) and we manually set it with the right value (e.g., dev, qa, prod). `local` is the default value and the one in use when the app runs on a developer's machine.

These are my use cases:

| debug | env           | expected  |
| ------- | ----------- | ----------- |
| true     | local         | read-only |
| true     | non-local | read-only |
| false    | local         | write         |
| false    | non-local | read-only |

I found out that `datasources.default.readOnly` can be set to true for this, but I also noticed that this property must be set as early in the `ApplicationContext` creation as possible. Otherwise, it will be too late, and the DB connection will have the default value (i.e., false).

I tried to use a `@Factory` together with a `@Singleton` to set the property to either true or false depending on the `environment` value, but it seems that by the time the singleton is created, it's already too late in the init process because it doesn't work (the property is added to the `Environment`, but the Connection Pool is already created by then).

An approach I've seen that may work is:

```yaml
debug: false
datasources:
  default:
    readOnly: ${debug}
```

But the downside is that I need to set the `datasources.*` in the properties, which may defeat my goal of reducing as much as possible the human error (developers shouldn't be able to commit changes from the app running locally but pointing to a Dev DB, for instance). Maybe I'm overthinking/overengineering the process, but I'm trying to make the debug experience as seamless and safer as possible.

I also tried implementing `PropertySource` (specifically, I extended the `MapPropertySource` class), but unfortunately, it seems that Micronaut doesn't automatically look up for `PropertySource` beans when starting up, and my custom bean was ignored.

I know it can be manually registered using the Micronaut builder used in the Main class, but I would like to know if there is another way of implementing this. I wouldn't like to use this approach since it can't be enforced (someone can change the code), and, more importantly, it's not testable. I would like to use an automatic "hook" or "standard way" so that the context created by `@MicronautTest` is the same that the one created by `Micronaut.run(Main.class, args)` so that I can create tests to enforce that this behavior is not altered (at least, without any good reason).

Can you tell me if this is possible to implement in a way I look for? Thanks in advance.

**Comment by Nahuel92** [comment]
~~Please don't download anything from the user above, it may be malware. I can't delete that comment, but I have reported & blocked the user.
dstepanov hi, can you please delete the previous comment? Sorry for citing you here, but I don't know how to help to prevent people from downloading that file. Thanks!!~~

Previous comment was deleted by GitHub in response to my report on a user sharing links to probably malware. Sorry for polluting the comment section.

**Comment by graemerocher** [comment, marked as answer]
the way you would do this is to write a `BeanCreatedEventListener<DatasourceConfiguration>` that called `setReadOnly(true)` but in the development or test environments

**Reply by Nahuel92** [reply]
Thanks Graeme!! That helped me to accomplish what I wanted to.

Thank you :)

**Reply by Nahuel92** [reply]
Hi @graemerocher, hey I have a quick question about the approach you shared. It works well, unless repository code runs in another thread (for example, when using a thread pool or a virtual thread executor).

In my mind, I have the idea that a Datasource/connection pool is created once and reused and share between threads. Can you confirm this is the case? If so, then I can share an example app where this behavior can be observed. Thanks.
